webrtc-stats.html

```diff
@@ -2218,6 +2220,40 @@ <h2>
                   formats have ways to vary the quantizer value within the frame.
                 </p>
               </dd>
+              <dt>
+                <dfn>psnrSum</dfn> of type <span class="idlMemberType">
+                record&lt;DOMString, double&gt;</span>
+              </dt>
+              <dd>
+                <p>
+                  MUST NOT [= map/exist =] for audio.
+                  The cumulative sum of the PSNR values of frames encoded by this sender.
+                  The record includes values for the "y", "u" and "v" components.
+                  The count of measurements is in {{psnrMeasurements}}.
+                </p>
+                <p>
+                  PSNR is defined in [[ISO-29170-1:2017]].
+                </p>
+                <p class="note">
+		  PSNR metrics should primarily be used as a basis for statistical analysis rather
+		  than be used as an absolute truth on a per-frame basis.
+                </p>
+              </dd>
+              <dt>
+                <dfn>psnrMeasurements</dfn> of type <span class="idlMemberType">unsigned long long
+                </span>
+              </dt>
+              <dd>
+                <p>
+                  MUST NOT [= map/exist =] for audio.
+                  The number of times PSNR was measured. The components of {{psnrSum}} are
+                  aggregated with this measurement.
+                </p>
+                <p>
+                  The PSNR is defined in [[ISO-29170-1:2017]].
+                  The frequency of PSNR measurements is [=implementation-defined=].
```

Comment:
OK with me if for any reasonable value of X (I'm not married to any particular value, as long as we have one that lets us WPT test) 
```suggestion
                  The frequency of PSNR measurements is [=implementation-defined=],
                  but SHOULD be no less than every X seconds.
```

Reply:
A single value is sufficient for existence and can be used to write a WPT.

Reply:
The argument here is a WPT testing an interoperable floor on frequency, not just existence.

> I only see benefits in the spec defining the implementor playground.

I believe @youennf also sought some implementer guidance here so we don't have to reverse engineer interoperable behavior.

If we can't resolve this in the PR let's aim to discuss this again next meeting.

Reply:
I still have not seen an argument why the minimum frequency needs to be interoperable.
Also I assume you mean "intraoperable"?

Reply:
I mean [interoperable](https://www.w3.org/policies/process/#implementation-experience), as in websites written and tested in one browser won't break in another.

I gave an example above in https://github.com/w3c/webrtc-stats/pull/794#issuecomment-2691600502 showing this is JS-observable implementation-defined behavior. A note doesn't change that.

[§ 11.2. Specify completely and avoid ambiguity](https://w3ctag.github.io/design-principles/#avoid-ambiguity) emphasizes that every feature should be sufficiently specified so there is no room for unpredictable or diverging per-implementation decisions.

This suggests we need good reasons to _tolerate_ implementation-defined behavior, not the other way around.

I haven't heard any argument why implementations need this much flexibility to lower their frequency of measurement that we can't specify a floor at all. This strikes me as unreasonable.

Technically speaking, we'd need a MUST for WPT to test this. So I think asking for a SHOULD, which amounts to little more than minimal implementation guidance, is being quite generous.

Maybe we need two frequencies: one that it SHOULD be on average, and a floor that it MUST be higher than?

Reply:
We seem to be talking past each other. No-one's objected to implementation-defined behavior here. A SHOULD *is* implementation-defined behavior. A floor on implementation-defined behavior is still implementation-defined behavior.

I'm asking the reason we can't give implementer guidance, or put any boundaries on it.

Reply:
well, then take the implementers feedback: such a number is not required and a specification constraining without a solid data basis and experimentation (such as done by the paper) is not how things are done in video encoding.

Reply:
Specs generally don't dictate quality or performance of media output. But that would be like if we put bounds on returned psnr values themselves. This is not what we're talking about.

The shape of `getStats` requires websites take two or more samples to divide by a the time in between. This creates a potential interoperability issue with obtaining valid measurements over short periods of time for this stat specifically (even if this is not the intended use) if the minimum measurement interval varies significantly by implementation. No expertise on psnr is needed to observe this.

Understanding that delivering measurements can be costly, I'd like feedback on how a generous 15 second upper cap on measurement interval would impose on implementers. The number is inherently arbitrary as it is picked to be outside the bounds of reasonable implementations.

If we can't agree on a number, some guidance to implementers and authors that would lead them to correctly guess measurement cadence as suggested in this thread (~every second, rarely ever more than every 5) would also do the job I think.

Reply:
How do you propose to satisfy this arbitrary when switching from an encoder that exposes PSNR to one that does not?

Reply:
Adding more unfounded numbers (unless you have run large scale experiments?) is not going to help.
1 second is quite likely to be an issue for high-resolution screen sharing.

Going once

Reply:
Edit: ^ s/every second/every few seconds/

Reply:
If situations arise where implementations fall outside these bounds we can always revisit.

Reply:
Since a vague soft limit doesn't avoid the "numerator is zero" what problem does this solve?
What is the behavior when the track was removed via replaceTrack? Follow-up question in #619...

Reply:
> Since a vague soft limit doesn't avoid the "numerator is zero" what problem does this solve?

@youennf does it address your concern in https://github.com/w3c/webrtc-stats/pull/794#issuecomment-2721716590?

> What is the behavior when the track was removed via replaceTrack?

The encoding frame rate drops to zero, which is lower than 1/15, and no PSNR measurements happen.

Reply:
How's this?
```suggestion
                  The frequency of PSNR measurements is [=implementation-defined=].
                </p>
                <p class="note">
                  User agents are encouraged to measure PSNR at a fixed frequency
                  for each encoder implementation, and as frequently as possibly as
                  long as it does not introduce a noticeable performance penalty.
```

Reply:
LGTM